misc-utils/lastlog2.c

```diff
@@ -0,0 +1,332 @@
+     notice, this list of conditions and the following disclaimer in the
+     documentation and/or other materials provided with the distribution.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+  POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <pwd.h>
+#include <time.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <getopt.h>
+#include <limits.h>
+
+#include "nls.h"
+#include "c.h"
+#include "strutils.h"
+#include "lastlog2.h"
+
+static char *lastlog2_path = LL2_DEFAULT_DATABASE;
+
+static int bflg = 0;
+static time_t b_days = 0;
+static int tflg = 0;
+static time_t t_days = 0;
+static int sflg = 0;
+
+static int
+print_entry (const char *user, int64_t ll_time,
+	     const char *tty, const char *rhost,
+	     const char *pam_service)
+{
+	static int once = 0;
+	char *datep;
+	struct tm *tm, tm_buf;
+	char datetime[80];
+	/* IPv6 address is at maximum 39 characters.
+	   But for LL-addresses (fe80+only) the interface should be set,
+	   so LL-address + % + IFNAMSIZ. */
+	const int maxIPv6Addrlen = 42;
+
+	/* Print only if older than b days */
+	if (bflg && ((time (NULL) - ll_time) < b_days))
+		return 0;
+
+	/* Print only if newer than t days */
+	if (tflg && ((time (NULL) - ll_time) > t_days))
+		return 0;
+        /* this is necessary if you compile this on architectures with
+           a 32bit time_t type. */
+        time_t t_time = ll_time;
+        tm = localtime_r (&t_time, &tm_buf);
+	if (tm == NULL)
+		datep = "(unknown)";
+	else {
+		strftime (datetime, sizeof (datetime), "%a %b %e %H:%M:%S %z %Y", tm);
+		datep = datetime;
+	}
+
+	if (ll_time == 0)
+		datep = "**Never logged in**";
+
+	if (!once) {
+		printf ("Username         Port     From%*s Latest%*s%s\n",
+			maxIPv6Addrlen-4, " ",
+			sflg?(int)strlen (datep)-5:0, " ", sflg?"Service":"");
+		once = 1;
+	}
+	printf ("%-16s %-8.8s %*s %s%*s%s\n", user, tty ? tty : "",
+		-maxIPv6Addrlen, rhost ? rhost : "", datep,
+		sflg?31-(int)strlen(datep):0, " ", sflg?(pam_service?pam_service:""):"");
+
+	return 0;
+}
+
+static void
+usage (int retval)
+{
+	FILE *output = (retval != EXIT_SUCCESS) ? stderr : stdout;
+
+	fprintf (output, _("Usage: lastlog2 [options]\n\nOptions:\n"));
```

Comment:
Here, we need a little bit to convert it to the util-linux way :-)
```
fputs(USAGE_HEADER, out);
```
See, for example, the recently added misc-utils/waitpid.c.

Reply:
Yes, really nice. I have adapted it.